faculty/clients/api.py

```diff
@@ -0,0 +1,635 @@
+            api_type=APIType.PLUMBER,
+            working_directory=working_directory,
+            script_name=script_name,
+        )
+
+    @staticmethod
+    def script_definition(working_directory, script_name):
+        return ScriptDefinition(
+            api_type=APIType.SCRIPT,
+            working_directory=working_directory,
+            script_name=script_name,
+        )
+
+    def _get_current_api_definition(self, project_id, api_id):
+        current_api = self.get(project_id, api_id)
+        current_api = _APISchema().dump(current_api)
+        return current_api
+
+    def get(self, project_id, api_id):
+        return self._get(
+            "/project/{}/api/{}".format(project_id, api_id), _APISchema()
+        )
+
+    def update(
+        self,
+        project_id,
+        api_id,
+        environment_ids=None,
+        api_definition=None,
+        server_type=None,
+        current_api=None,
+    ):
+        if not current_api:
+            current_api = self._get_current_api_definition(project_id, api_id)
+        payload = {}
+        payload["defaultServerSize"] = current_api["defaultServerSize"]
+        payload["environmentIds"] = current_api["environmentIds"]
+        payload["definition"] = current_api["definition"]
+        if server_type:
+            payload["defaultServerSize"] = server_type
+        if environment_ids:
+            payload["environmentIds"] = environment_ids
+
+        if api_definition:
+            payload["definition"] = api_definition
+        endpoint = "/project/{}/api/{}/definition".format(project_id, api_id)
+        return self._put(endpoint, _APISchema(), json=payload)
+
+    def delete(self, project_id, api_id):
+        return self._delete_raw(
+            "/project/{}/api/{}".format(project_id, api_id)
+        )
+
+    def list(self, project_id):
+        """List APIs in a project.
+
+        Parameters
+        ----------
+        project_id : uuid.UUID
+            The project to list APIs in.
+        name : str, optional
+            If provided, only return APIs with this name.
+
+        Returns
+        -------
+        List[API]
+            The matching APIs.
+        """
+        endpoint = "/project/{}/api".format(project_id)
+        return self._get(endpoint, _APISchema(many=True))
+
+    def create_production_key(self, project_id, api_id, label=None):
+        """Create a production key for a given API.
+
+        Parameters
+        ----------
+        project_id : uuid.UUID
+            The project where the API resides.
+        api_id : uuid.UUID
+            The API to create a new key in.
+        label : str, optional
+            If provided, set this value as the new key's label.
+
+        Returns
+        -------
+        APIKey
+            The newly created key.
+        """
+        endpoint = "/project/{}/api/{}/key".format(project_id, api_id)
+        # TODO: maybe more sensible default, the UI doesn't allow empty name but the api accepts this
```

Comment:
Does the API accept a missing (as opposed to empty) label? If so, I'd suggest not setting a default here, if not, I'd suggest forcing the caller to provide one.

Reply:
Missing value is not accepted, will go with the second option of making it a mandatory input.